atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraServersConfigs.java

```diff
@@ -0,0 +1,222 @@
+        <T> T visit(Visitor<T> visitor);
+
+        int numberOfHosts();
+
+        void check();
+    }
+
+    @Value.Immutable
+    @JsonDeserialize(as = ImmutableDefaultConfig.class)
+    @JsonSerialize(as = ImmutableDefaultConfig.class)
+    @JsonTypeName(DefaultConfig.TYPE)
+    public abstract static class DefaultConfig implements CassandraServersConfig {
+        public static final String TYPE = "default";
+
+        @Override
+        public final <T> T visit(Visitor<T> visitor) {
+            return visitor.apply(thrift(), Optional.empty());
+        }
+
+        @JsonValue
+        public abstract Set<InetSocketAddress> thrift();
+
+        @Override
+        public int numberOfHosts() {
+            return thrift().size();
+        }
+
+        @Override
+        public void check() {
+            Preconditions.checkState(!thrift().isEmpty(), "'servers' must have at least one entry");
+            for (InetSocketAddress addr : thrift()) {
+                Preconditions.checkState(addr.getPort() > 0, "each server must specify a port ([host]:[port])");
+            }
+        }
+    }
+
+
+    @Value.Immutable
+    @JsonDeserialize(as = ImmutableThriftOnlyConfig.class)
+    @JsonSerialize(as = ImmutableThriftOnlyConfig.class)
+    @JsonTypeName(ThriftOnlyConfig.TYPE)
+    public abstract static class ThriftOnlyConfig implements CassandraServersConfig {
+        public static final String TYPE = "thriftOnly";
+
+        @Override
+        public final <T> T visit(Visitor<T> visitor) {
+            return visitor.apply(thrift(), Optional.empty());
+        }
+
+        @Override
+        public final int numberOfHosts() {
+            return thrift().size();
+        }
+
+        @Override
+        public final void check() {
+            Preconditions.checkState(!thrift().isEmpty(), "'servers' must have at least one entry");
+            for (InetSocketAddress addr : thrift()) {
+                Preconditions.checkState(addr.getPort() > 0, "each server must specify a port ([host]:[port])");
+            }
+        }
+
+        @JsonProperty
+        public abstract Set<InetSocketAddress> thrift();
+
+        @JsonProperty
+        @Value.Default
+        String type() {
+            return TYPE;
+        }
+    }
+
+    @JsonDeserialize(as = ImmutableCqlCapableConfig.class)
+    @JsonSerialize(as = ImmutableCqlCapableConfig.class)
+    @JsonTypeName(CqlCapableConfig.TYPE)
+    @Value.Immutable
+    public abstract static class CqlCapableConfig implements CassandraServersConfig {
+        public static final String TYPE = "cqlCapable";
+
+        @Value.Immutable
+        @JsonDeserialize(as = ImmutableCqlCapableServer.class)
+        @JsonSerialize(as = ImmutableCqlCapableServer.class)
+        interface CqlCapableServer {
+
+            @JsonProperty
+            String hostname();
+            @JsonProperty
+            Integer thriftPort();
+
+            @JsonProperty
+            Integer cqlPort();
```

Comment:
Is there any reason we went for the reference-types here?

Reply:
No reason, changed to value types.